test/replication-luatest/gh_6033_box_promote_demote_test.lua

```diff
@@ -0,0 +1,365 @@
+        s:create_index('pk')
+        require('fiber').create(s.replace, s, {1})
+    end)
+    g.server_1:wait_wal_write_count(wal_write_count + 1)
+
+    g.server_3:drop()
+
+    local ok, err = g.server_1:exec(function()
+        box.cfg{replication_synchro_timeout = 0.1}
+        return pcall(box.ctl.promote)
+    end)
+
+    g.server_2:exec(function()
+        box.error.injection.set('ERRINJ_WAL_DELAY', false)
+    end)
+    luatest.assert(
+        not ok and err.code == box.error.QUORUM_WAIT,
+        'wait quorum failure not handled')
+end
+
+g.after_test('test_fail_limbo_acked_promote', function(g)
+    g.server_3:drop()
+    g.server_3 = nil
+end)
+
+-- Test interfering demotion for election_mode = 'off'
+-- while in WAL delay
+g.before_test('test_wal_interfering_demote', function(g)
+    g.server_1:exec(function()
+        box.ctl.promote()
+    end)
+    g.server_1:wait_synchro_queue_owner()
+    g.server_2:wait_synchro_queue_owner(g.server_1:instance_id())
+end)
+
+g.test_wal_interfering_demote = function(g)
+    local wal_write_count = get_wal_write_count(g.server_1)
+    local raft_term = get_raft_term(g.server_1)
+
+    g.server_1:exec(function()
+        box.error.injection.set('ERRINJ_WAL_DELAY', true)
+    end)
+
+    g.server_2:exec(function()
+        box.ctl.promote()
+    end)
+    g.server_2:wait_synchro_queue_owner()
+    wait_promote(g.server_1, wal_write_count, raft_term)
+
+    local ok, err = g.server_1:exec(function()
+        local f = require('fiber').new(box.ctl.demote); f:set_joinable(true)
+        box.error.injection.set('ERRINJ_WAL_DELAY', false)
+        return f:join()
+    end)
+    luatest.assert(
+        not ok and err.code == box.error.INTERFERING_PROMOTE,
+        'interfering demote not handled')
+end
+
+-- Test interfering demotion for election_mode = 'off'
+-- while in txn_limbo delay
+g.test_limbo_empty_interfering_demote = function(g)
+    local wal_write_count = get_wal_write_count(g.server_1)
+    local raft_term = get_raft_term(g.server_1)
+
+    local f = g.server_1:exec(function()
+        box.error.injection.set('ERRINJ_TXN_LIMBO_EMPTY_DELAY', true)
+        local f = require('fiber').new(box.ctl.promote); f:set_joinable(true)
+        return f:id()
+    end)
+
+    g.server_2:exec(function()
+        box.ctl.promote()
+    end)
+    g.server_2:wait_synchro_queue_owner()
+    wait_promote(g.server_1, wal_write_count, raft_term)
+
+    local ok, err = g.server_1:exec(function(f)
+        box.error.injection.set('ERRINJ_TXN_LIMBO_EMPTY_DELAY', false)
+        return require('fiber').find(f):join()
+    end, {f})
+    luatest.assert(
+        not ok and err.code == box.error.INTERFERING_PROMOTE,
+        'interfering demote not handled')
+end
+
+-- Test that box_demote will return 0 when demoting raft leader
+g.before_test('test_raft_leader_demote', function(g)
+    g.server_1:box_config({election_mode = 'manual'})
+    make_leader(g.server_1)
+end)
+
+g.test_raft_leader_demote = function(g)
```

Comment:
I don't understand why is this test needed. Wasn't this case covered before your test?
Maybe you meant to demote server_2 and check that there is no error?